### ISSUE TITLE
Global regular matches generate invalid pointers

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8090]
Migrated from [rt.perl.org#37038](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=37038) (status was 'resolved')

Searchable as RT37038$

**Comment by p5pRT** [comment]
### From tgirmann@web.de

#### Created by tgirmann@web.de



\(I'm sending this message again using a mail client as it appears
to me that it wasn't delivered by sendmail\)

When&#8203;:
\* doing a global regular match
\* using $1 $2 etc\. to retrieve extracted substrings
\* the original string does not exist any longer

then $1 $2 etc\. is pointing to an unallocated piece of memory\.
This way it's possible to return Perl's internal data structures
if the piece of memory got re\-used and possibly also cause Perl
to crash if that piece of memory has been returned to the
operating system\.

Here's a simple example to confirm the bug \(tested under Windows
NT / ActivePerl 5\.8\.7 and Linux 2\.4 / Perl 5\.8\.0\)&#8203;:

use strict;
use warnings;
my $s = "abcd";
$s =~ /\(\.\.\)\(\.\.\)/g;
$s = $1;
$s = $2;
print "$s\\n";


The statement "$s = $1" causes Perl to re\-use the string "abcd"
for storing the first half \("ab"\) and add a terminating zero\.
$2 still points to the original second half but the character
"c" got overwritten neanwhile\.

So if the bug is present you will get " c" or just "c" printed
out instead of the expected "cd"\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.8.7:

Configured by builder at Mon Jun  6 13:36:05 2005.

Summary of my perl5 (revision 5 version 8 subversion 7) configuration:
  Platform:
    osname=MSWin32, osvers=5.0, archname=MSWin32-x86-multi-thread
    uname=''
    config_args='undef'
    hint=recommended, useposix=true, d_sigaction=undef
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=de
fine
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cl', ccflags ='-nologo -Gf -W3 -MD -Zi -DNDEBUG -O1 -DWIN32 -D_CONSOLE -DNO_STRICT -DHAVE_DES_FCRYPT -DBUILT_BY_ACTIVESTATE -DNO_HASH_SEED -DUSE_SITECUSTOMIZE -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -DPERL_MSVCRT_READFIX',
    optimize='-MD -Zi -DNDEBUG -O1',
    cppflags='-DWIN32'
    ccversion='12.00.8804', gccversion='', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=undef, longlongsize=8, d_longdbl=define, longdblsize=10
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='__int64', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='link', ldflags ='-nologo -nodefaultlib -debug -opt:ref,icf  -libpath:"C:\Perl\lib\CORE"  -machine:x86'
    libpth=\lib
    libs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib msvcrt.lib
    perllibs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib msvcrt.lib
    libc=msvcrt.lib, so=dll, useshrplib=yes, libperl=perl58.lib
    gnulibc_version='undef'
  Dynamic Linking:
    dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib -debug -opt:ref,icf  -libpath:"C:\Perl\lib\CORE"  -machine:x86'

Locally applied patches:
    ACTIVEPERL_LOCAL_PATCHES_ENTRY
    #  if !defined(PERL_DARWIN)
    Iin_load_module moved for compatibility with build 806
    #  endif
    #  if defined(__hpux)
    Avoid signal flag SA_RESTART for older versions of HP-UX
    #  endif
    PerlEx hacks for CGI::Carp
    Less verbose ExtUtils::Install and Pod::Find
    instmodsh upgraded from ExtUtils-MakeMaker-6.25
    24699 ICMP_UNREACHABLE handling in Net::Ping
    21540 Fix backward-compatibility issues in if.pm


@INC for perl v5.8.7:
    C:/Perl/lib
    C:/Perl/site/lib
    .


Environment for perl v5.8.7:
    HOME (unset)
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=c:\programme\imagemagick-6.2.4-q16;C:\Perl\bin\;C:\WINDOWS\system32;C:\WINDOWS;C:\Programme\UltraEdit;C:\Programme\Microsoft SQL Server\80\Tools\BINN;C:\Tools;C:\Programme\Gemeinsame Dateien\GTK\2.0\bin;C:\Programme\Mts
    PERL_BADLANG (unset)
    SHELL (unset)



```
</details>

**Comment by p5pRT** [comment]
### From @demerphq

On 8/31/05\, via RT Thilo Girmann \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> \# New Ticket Created by  Thilo Girmann
> \# Please include the string&#8203;:  \[perl \#37038\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=37038 >
> 
> 
> 
> This is a bug report for perl from tgirmann@&#8203;web\.de\,
> generated with the help of perlbug 1\.35 running under perl v5\.8\.7\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> \(I'm sending this message again using a mail client as it appears
> to me that it wasn't delivered by sendmail\)
> 
> When&#8203;:
> \* doing a global regular match
> \* using $1 $2 etc\. to retrieve extracted substrings
> \* the original string does not exist any longer
> 
> then $1 $2 etc\. is pointing to an unallocated piece of memory\.
> This way it's possible to return Perl's internal data structures
> if the piece of memory got re\-used and possibly also cause Perl
> to crash if that piece of memory has been returned to the
> operating system\.
> 
> Here's a simple example to confirm the bug \(tested under Windows
> NT / ActivePerl 5\.8\.7 and Linux 2\.4 / Perl 5\.8\.0\)&#8203;:
> 
> use strict;
> use warnings;
> my $s = "abcd";
> $s =~ /\(\.\.\)\(\.\.\)/g;
> $s = $1;
> $s = $2;
> print "$s\\n";
> 
> 
> The statement "$s = $1" causes Perl to re\-use the string "abcd"
> for storing the first half \("ab"\) and add a terminating zero\.
> $2 still points to the original second half but the character
> "c" got overwritten neanwhile\.
> 
> So if the bug is present you will get " c" or just "c" printed
> out instead of the expected "cd"\.

I can confirm this behaviour is in 5\.8\.6 and 5\.6\.1 on Win32\.

yves

\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @nwc10

On Wed\, Aug 31\, 2005 at 05&#8203;:32&#8203;:25AM \-0700\, Thilo Girmann wrote&#8203;:

> The statement "$s = $1" causes Perl to re\-use the string "abcd"
> for storing the first half \("ab"\) and add a terminating zero\.
> $2 still points to the original second half but the character
> "c" got overwritten neanwhile\.
> 
> So if the bug is present you will get " c" or just "c" printed
> out instead of the expected "cd"\.

I've added a TODO regression test for this in t/op/pat\.t\, but I'm not sure
how to fix this\, without creating a speed hit by copying the matched string
every time\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @nwc10

On Tue\, Sep 06\, 2005 at 07&#8203;:14&#8203;:27PM \+0100\, Nicholas Clark wrote&#8203;:
> On Wed\, Aug 31\, 2005 at 05&#8203;:32&#8203;:25AM \-0700\, Thilo Girmann wrote&#8203;:
> 
> > The statement "$s = $1" causes Perl to re\-use the string "abcd"
> > for storing the first half \("ab"\) and add a terminating zero\.
> > $2 still points to the original second half but the character
> > "c" got overwritten neanwhile\.
> > 
> > So if the bug is present you will get " c" or just "c" printed
> > out instead of the expected "cd"\.
> 
> I've added a TODO regression test for this in t/op/pat\.t\, but I'm not sure
> how to fix this\, without creating a speed hit by copying the matched string
> every time\.

Specifically I don't think that we can spot the pattern of assigning to
the match variable via static analysis \(consider references\)\, so whatever
we do would have to be runtime\. Therefore if we aren't copying\, we need to
fake it as if we were\, which means some sort of copy on write solution\.
Albeit special case\, as we can probably hide sufficient data inside the
regexp rather than the SV\. But I can't work out a good way to trigger the
"on write" part\, suitable for maint\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @cpansprout

I was trying to force the pre\-match copy to do COW\, for the sake of fixing //g without slowing it down \(bugs \#37038 and \#46563\)\.

I ran into some interesting ‘panic&#8203;: bad free’ errors\, and I can’t figure out why\.

I suspect there is still some problem with sv\_sethek\.

See the attached patch \(where I’ve changed the error to an abort\, so that gdb will stop there\)\.

It fails for me in the middle of autodoc\.pl\, but the line number it stops on doesn’t seem particularly informative\.

If you comment out the indicated sv\_sethek line\, the problem goes away\, but so does the speed\-up I was trying to get\.

Actually\, if I change the offending line to

&nbsp;        sv\_setsv\_nomg\(sv\, sv\_2mortal\(newSVhek\(hek\)\)\);

the problem persists\.  So it seems to be related to shared heks in general\.

Am I doing something obviously wrong?

<details><summary>Inline Patch</summary>

```diff
diff --git a/dump.c b/dump.c
index 2c635de..b5ef034 100644
--- a/dump.c
+++ b/dump.c
@@ -2063,10 +2063,8 @@ Perl_do_sv_dump(pTHX_ I32 level, PerlIO *file, SV *sv, I32 nest, I32 maxnest, bo
 				PTR2UV(r->pprivate));
 	    Perl_dump_indent(aTHX_ level, file, "  OFFS = 0x%"UVxf"\n",
 				PTR2UV(r->offs));
-#ifdef PERL_OLD_COPY_ON_WRITE
 	    Perl_dump_indent(aTHX_ level, file, "  SAVED_COPY = 0x%"UVxf"\n",
 				PTR2UV(r->saved_copy));
-#endif
 	}
 	break;
     }
diff --git a/regcomp.c b/regcomp.c
index 4669c27..a9da541 100644
--- a/regcomp.c
+++ b/regcomp.c
@@ -11757,9 +11757,7 @@ Perl_pregfree2(pTHX_ REGEXP *rx)
 	Safefree(r->substrs);
     }
     RX_MATCH_COPY_FREE(rx);
-#ifdef PERL_OLD_COPY_ON_WRITE
     SvREFCNT_dec(r->saved_copy);
-#endif
     Safefree(r->offs);
 }
 
@@ -11820,9 +11818,7 @@ Perl_reg_temp_copy (pTHX_ REGEXP *ret_x, REGEXP *rx)
 	   anchored or float namesakes, and don't hold a second reference.  */
     }
     RX_MATCH_COPIED_off(ret_x);
-#ifdef PERL_OLD_COPY_ON_WRITE
     ret->saved_copy = NULL;
-#endif
     ret->mother_re = rx;
     
     return ret_x;
@@ -12040,9 +12036,7 @@ Perl_re_dup_guts(pTHX_ const REGEXP *sstr, REGEXP *dstr, CLONE_PARAMS *param)
 	ret->subbeg  = SAVEPVN(ret->subbeg, ret->sublen);
     else
 	ret->subbeg = NULL;
-#ifdef PERL_OLD_COPY_ON_WRITE
     ret->saved_copy = NULL;
-#endif
 
     if (ret->mother_re) {
 	if (SvPVX_const(dstr) == SvPVX_const(ret->mother_re)) {
diff --git a/regexec.c b/regexec.c
index 44d7b96..259446d 100644
--- a/regexec.c
+++ b/regexec.c
@@ -2510,9 +2510,28 @@ got_it:
 		prog->saved_copy = sv_setsv_cow(prog->saved_copy, sv);
 		prog->subbeg = (char *)SvPVX_const(prog->saved_copy);
 		assert (SvPOKp(prog->saved_copy));
-	    } else
+#else
+	    if (SvIsCOW(sv)) {
+		if (!prog->saved_copy) prog->saved_copy = newSV(0);
+		sv_setsv_nomg(prog->saved_copy, sv);
+		prog->subbeg = (char *)SvPVX_const(prog->saved_copy);
+	    }
+	    else if ((STRLEN)i == SvCUR(sv)
+		  && (SvFLAGS(sv) & CAN_COW_MASK) == CAN_COW_FLAGS) {
+		/* Force a COW.  Creating a shared string makes exactly
+		   one copy, which we were going to make anyway.  If this
+		   string will be subjected to multiple matches, this will
+		   save us making another copy for each match. */
+		HEK *hek = share_hek(strbeg, SvUTF8(sv) ? -i : i, 0);
+	/* comment out the following line to avoid the error: */
+		sv_sethek(sv, hek);
+		if (prog->saved_copy) sv_sethek(prog->saved_copy, hek);
+		else prog->saved_copy = newSVhek(hek);
+		unshare_hek(hek);
+		prog->subbeg = (char *)SvPVX_const(prog->saved_copy);
 #endif
-	    {
+	    }
+	    else {
 		RX_MATCH_COPIED_on(rx);
 		s = savepvn(strbeg, i);
 		prog->subbeg = s;
diff --git a/regexp.h b/regexp.h
index 5ee6448..43186a5 100644
--- a/regexp.h
+++ b/regexp.h
@@ -44,12 +44,6 @@ struct reg_substr_data {
     struct reg_substr_datum data[3];	/* Actual array */
 };
 
-#ifdef PERL_OLD_COPY_ON_WRITE
-#define SV_SAVED_COPY   SV *saved_copy; /* If non-NULL, SV which is COW from original */
-#else
-#define SV_SAVED_COPY
-#endif
-
 typedef struct regexp_paren_pair {
     I32 start;
     I32 end;
@@ -94,7 +88,7 @@ typedef struct regexp_paren_pair {
 	regexp_paren_pair *offs;					\
 	/* saved or original string so \digit works forever. */		\
 	char *subbeg;							\
-	SV_SAVED_COPY	/* If non-NULL, SV which is COW from original */\
+	SV *saved_copy;	/* If non-NULL, SV which is COW from original */\
 	I32 sublen;	/* Length of string pointed by subbeg */	\
 	/* Information about the match that isn't often used */		\
 	/* offset from wrapped to the start of precomp */		\
diff --git a/util.c b/util.c
index 9d7683d..ef5f1a0 100644
--- a/util.c
+++ b/util.c
@@ -265,6 +265,7 @@ Perl_safesysfree(Malloc_t where)
 	    }
 	    if (!(header->next) || header->next->prev != header
 		|| header->prev->next != header) {
+abort();
 		Perl_croak_nocontext("panic: bad free");
 	    }
 	    /* Unlink us from the chain.  */
```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

Fixed by a41aa44d9dc\.
\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Fixed by a41aa44d9dc\.
\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
